### DISCUSSION TITLE
Decoupling WinningPoSt and WindowPoSt Process from lotus-miner

### DISCUSSION THREAD

**Discussion by jennijuju** [filecoin-project/lotus#6943] [Ideas] [answered]
**Author: @siriusmz**

## Design summary

![](../images/post_decoupling.jpg)

- Miner gets sector storage info and sector path for PoSt worker.
- PoSt worker gets VanillaPoSt by HTTP from remote workers.
- Miner schedules window PoSt workers to calculate WindowPost in parallel for batch. 
- Miner schedules PoSt workers based on GPU . 
- Multiple window workers can be deployed according to the size of the cluster.

### Add interface

- Add filecoin-ffi interface

```
// Sorting and de-duplication sector info
func DeDuplicatePrivateSectorInfo(ctx context.Context, privsectors *SortedPrivateSectorInfo) (SortedPrivateSectorInfo, error)

// Get sector info from start to end
func SplitSortedPrivateSectorInfo(ctx context.Context, privsectors SortedPrivateSectorInfo, start int, end int) (SortedPrivateSectorInfo, error)

// Get the number of sectors that need to be certified for each partition of the window
func GetPartitionSectorCount(ctx context.Context, proofType int, repslen int) (int, error)

// Get vanilla PoSt params of partition
func GetPartitionVanillaParams(ctx context.Context,proofType abi.RegisteredPoStProof,sectorsLen uint64,) (partitionVanillaParams string, err error)

// Get vanilla PoSt params of sector
func GetSectorVanillaParams(ctx context.Context, offset int, partitionVanillaParams string) （sectorVanillaParams string, err error)

// Generate window vanilla PoSt for the given sectors which compose a partition
func GenerateWindowPoStVanilla(ctx context.Context,minerID abi.ActorID,privsector PrivateSectorInfo,randomness abi.PoStRandomness,sectorVanillaParams string) (string, error)

// Generate window PoSt with generated window vanilla PoSt result
func GenerateWindowPoStWithVanillas(ctx context.Context,minerID abi.ActorID,privsectors SortedPrivateSectorInfo,randomness abi.PoStRandomness,vanillas []string) ([]proof5.PoStProof, error)

// Generate winning vanilla PoSt for the given sector
func GenerateWinningPoStVanilla(ctx context.Context,minerID abi.ActorID,privsector PrivateSectorInfo,randomness abi.PoStRandomness) (vanilla string, err error)

// Generate winning PoSt with generated winning vanilla PoSt result
func GenerateWinningPoStWithVanillas(ctx context.Context,minerID abi.ActorID,privsectors SortedPrivateSectorInfo,randomness abi.PoStRandomness,vanillas []string) ([]proof5.PoStProof, error)
```

- Add index interface
>Used to obtain sector's URL

```
StorageGetUrl(ctx context.Context, s abi.SectorID, ft storiface.SectorFileType) (string, error)
```

- Extend ffiwraper.storage interface

> Used to support storage to call ffi-wrapper interface to obtain vanilla PoSt result

```
type WorkerProver interface {
	// worker_local generate window vanilla PoSt
    GenerateWindowPoStWithVanilla(ctx context.Context, mid abi.ActorID, privsectors SortedPrivateSectorInfo, randomness abi.PoStRandomness, vanillas []string) ([]proof.PoStProof, error)
    
    // worker_local generate winning vanilla PoSt
    GenerateWinningPoStWithVanilla(ctx context.Context, mid abi.ActorID, privsectors SortedPrivateSectorInfo, randomness abi.PoStRandomness, vanillas []string) ([]proof.PoStProof, error)
    
    // worker_local get vanilla PoSt params
    GetSectorVanillaParams(ctx context.Context, offset int, partitionVanillaParams string) string
}
```

- Extend SectorManager interface

> Support the manager to call the ffiwrapper interface to obtain the basic information that needed when generating PoSt, including sector information, vanilla PoSt params and the number of sectors in each partition.

```
type MinerProver interface {
	// SectorManager gets sorted and de-duplication sectors info.
    PubSectorToPriv(ctx context.Context, mid abi.ActorID, sectorInfo []proof5.SectorInfo, faults []abi.SectorNumber, rpt func(abi.RegisteredSealProof) (abi.RegisteredPoStProof, error)) (SortedPrivateSectorInfo, []abi.SectorID, func(), error)
    
    // SectorManager splits the sector sets according to the number of sectors in each partition.
    SplitSortedPrivateSectorInfo(ctx context.Context, privsectors SortedPrivateSectorInfo, off int, end int) (SortedPrivateSectorInfo, error)
    
    GetPartitionSectorsCount(ctx context.Context, prooftype abi.RegisteredPoStProof, sectorsLen int) (int, error)
    
    GetPartitionVanillaParams(ctx context.Context, proofType abi.RegisteredPoStProof, sectorsLen uint64) (string, error)
}
```

- Extend workerCalls interface

> Proof interface supported by PoSt worker

```
type WorkerCalls interface {
    GenerateWindowPoSt(ctx context.Context, mid abi.ActorID, privsectors SortedPrivateSectorInfo, partitionVanillaParams string, offset int, randomness abi.PoStRandomness) (WindowPoStResult, error)
    
    GenerateWinningPoSt(ctx context.Context, mid abi.ActorID, privsectors SortedPrivateSectorInfo, randomness abi.PoStRandomness) ([]proof.PoStProof, error)
}
```

- Add stores.Store interface

> Obtain the sector's vanilla PoSt from storage by HTTP

```
//Obtain the vanilla PoSt by HTTP
GenerateWindowPoStVanilla(ctx context.Context, minerID abi.ActorID, privsector *ffiwrapper.PrivateSectorInfo, vanillaParams string, randomness abi.PoStRandomness) (string, error)
GenerateWinningPoStVanilla(ctx context.Context, minerID abi.ActorID, privsector *ffiwrapper.PrivateSectorInfo, randomness abi.PoStRandomness) (string, error)
```

### PoSt worker scheduler

```
type poStScheduler struct {
    lk       sync.RWMutex
    // PoSt worker map
    workers  map[WorkerID]*workerHandle
    cond     *sync.Cond
    // PoSt type TTGenerateWindowPoSt or TTGenerateWinningPoSt
    postType sealtasks.TaskType
}
// add post worker
func (ps *poStScheduler) AddWorker(wid WorkerID, w *workerHandle) bool

// del post worker
func (ps *poStScheduler) delWorker(wid WorkerID) *workerHandle 

// determine whether there are workers that can be scheduled
func (ps *poStScheduler) CanSched(ctx context.Context) bool 

// schedule PoSt worker
func (ps *poStScheduler) SchedWorker(ctx context.Context, primary bool) (Worker, PoStDone, error) 

// determine whether the worker can handle the task
func (ps *poStScheduler) canHandleWorkers() (bool, []WorkerID)

// get resource
func (ps *poStScheduler) addResource(wid WorkerID)

// free resource
func (ps *poStScheduler) freeResource(wid WorkerID) 

// determine whether the worker is valid 
func (ps *poStScheduler) watch(wid WorkerID, worker *workerHandle) 

// clean up worker when lost connection
func (ps *poStScheduler) workerCleanup(wid WorkerID, w *workerHandle) 

// close PoSt worker scheduler
func (ps *poStScheduler) schedClose()

// get worker's status
func (ps *poStScheduler) WorkerStats(cb func(wid WorkerID, worker *workerHandle))
```

- The scheduling of window PoSt and winning PoSt is realized by instantiating PoSt scheduler.

```
type scheduler struct {
...
    // window scheduler
    windowPoStSched *poStScheduler
    
    // winning scheduler
    winningPoStSched *poStScheduler

...
}
```

- Through the above method, the window worker machine can be expanded to multiple.

### Compatible with original PoSt

```
func (m *Manager) GenerateWinningPoSt(ctx context.Context, minerID abi.ActorID, sectorInfo []proof.SectorInfo, randomness abi.PoStRandomness) ([]proof.PoStProof, error) {
    if !m.sched.winningPoStSched.CanSched(ctx) {
        return m.Prover.GenerateWinningPoSt(ctx, minerID, sectorInfo, randomness)
    }
    
    // Perform scheduling to calculate winning PoSt
    return m.generateWinningPoSt(ctx, minerID, sectorInfo, randomness)
}

func (m *Manager) GenerateWindowPoSt(ctx context.Context, minerID abi.ActorID, sectorInfo []proof.SectorInfo, randomness abi.PoStRandomness) (proof []proof.PoStProof, skipped []abi.SectorID, err error) {
    if !m.sched.windowPoStSched.CanSched(ctx) {
        return m.Prover.GenerateWindowPoSt(ctx, minerID, sectorInfo, randomness)
    }
    
	// Perform scheduling to calculate window PoSt
    return m.generateWindowPoSt(ctx, minerID, sectorInfo, randomness)
}
```

- The generateWindowPoSt function splits the batch to partitions and calculates them in parallel.

  

### Compute each batch in parallel

- Modify the logic of the runPoStCycle function to calculate in parallel between multiple batches

### Get vanilla PoSt

- http_handler implements generating vanilla PoSt interface

```
func (handler *FetchHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
    mux.HandleFunc("/remote/vanilla/window", handler.generateWindowPoStVanilla).Methods("POST")
    mux.HandleFunc("/remote/vanilla/winning", handler.generateWinningPoStVanilla).Methods("POST")
...
}
```

- The PoSt workers call the GenerateWindowPoStVanilla or GenerateWinningPoStVanilla interface which implemented by stores.Store  to obtain vanilla PoSt result from the sector's storage.

**Comment by jennijuju** [comment]
@magik6k @raulk I feel like a new design requirement for this would be -  it need to work with the `Subsystem` architecture introduced in v1.11.1. What do you think?

**Reply by dd45e640b42e6da7da96faee3996ef7c** [reply]
i am not raul or magic, but i see the same "problem".

**Reply by dd45e640b42e6da7da96faee3996ef7c** [reply]
one question i see when looking at the config vars for the subsystems: will they implement the same thing but via subsystems with, in a later stage, a N:M setup?

**Reply by siriusyim** [reply]
hey @jennijuju i think it can work with the Subsystem architecture introduced in v1.11.1

**Reply by siriusyim** [reply]
Now this task is coding and debugging on devnet

**Comment by siriusyim** [comment]
hey @jennijuju ，I am testing this post decoupling function. I will submit the code here temporarily. If there is no problem in the test, I will submit a PR: https://github.com/wusirdcenter/lotus/tree/feat/decoupling-post-worker
https://github.com/wusirdcenter/filecoin-ffi/tree/feat/decoupling-post-worker
https://github.com/wusirdcenter/rust-fil-proofs/tree/feat/decoupling-post-worker
https://github.com/wusirdcenter/rust-filecoin-proofs-api/tree/feat/decoupling-post-worker

**Reply by jennijuju** [reply]
How’s it going?

**Reply by siriusyim** [reply]
> How’s it going?
hi @jennijuju 
The test on devnet is passed, and it is currently being tested on the calibration network. After the calibnet test is passed, a PR will be submitted this week

**Reply by KIMQI001** [reply]
@siriusmz I always get this in devnet.
   WARN    advmgr  sector-storage/faults.go:95     CheckProvable Sector FAULT: generating vanilla proof    {"sector": {"ID":{"Miner":1008,"Number"
:15},"ProofType":6}, "sealed": "/app/zoe/test/miner-1/sealed/s-t01008-15", "cache": "/app/zoe/test/miner-1/cache/s-t01008-15", "err": "non-200 code: %!w(string=)", "errVerbosConnection closed: %!w(string=):\n    github.com/filecoin-project/lotus/extern/sector-storage/stores.(*Remote).GenerateSingleVanillaProof\n        /app/zoe/wn-rpc/extern/sector-storage/stores/remote.go:919"}

**Reply by siriusyim** [reply]
If the sector file is stored with miner as the entry, please try to modify .lotusminer/config.toml ListenAddress to "/ip4/0.0.0.0/tcp/2345/http"，and restart the lotus-miner

**Reply by siriusyim** [reply]
@KIMQI001 Now we are discussing with PL about the rectification of the rust interface, please do not go to the production environment

**Reply by KIMQI001** [reply]
@siriusmz okay,thx for reminding

**Reply by KIMQI001** [reply]
@siriusmz another question ,  how do I disable the wn|wd post of miner

**Reply by KIMQI001** [reply]
It always compute on miner's local-worker

**Reply by siriusyim** [reply]
If you stop wd｜wn worker, you can stop the worker, when you must ensure that all sectors use lotus-miner as the storage entry, otherwise the post cannot be completed

**Reply by siriusyim** [reply]
> It always compute on miner's local-worker
Do you mean that wn｜wd worker is started, post is still calculated on lotus-miner?

**Reply by KIMQI001** [reply]
I want let the miner as the storage entry , the remote worker as the post compute-unit.
But now the miner's local-worker always get the post-compute task...

**Reply by KIMQI001** [reply]
> > It always compute on miner's local-worker
> > Do you mean that wn｜wd worker is started, post is still calculated on lotus-miner?

yes

**Reply by siriusyim** [reply]
If the lotus-worker process is started on the miner machine, do not let him support wn｜wd, such as lotus-worker run --windowpost=false --winningpost=false

**Reply by KIMQI001** [reply]
it works perfectly 👍

**Reply by PerrywLiu** [reply]
> @KIMQI001 Now we are discussing with PL about the rectification of the rust interface, please do not go to the production environment

How is it going? Is it ready for production environment now？

**Reply by siriusyim** [reply]
> > @KIMQI001 Now we are discussing with PL about the rectification of the rust interface, please do not go to the production environment
> 
> How is it going? Is it ready for production environment now？

PR has been submitted, waiting for code approval and merge

**Reply by PerrywLiu** [reply]
Great ! I'm going to study and test it locally, Looking forward to an early merger~！

**Reply by dd45e640b42e6da7da96faee3996ef7c** [reply]
> If the lotus-worker process is started on the miner machine, do not let him support wn｜wd, such as lotus-worker run --windowpost=false --winningpost=false

will this be the case in the finished "product"? this needs to be possible

**Comment by jennijuju** [comment]
@siriusmz could you please confirm if the wdpost worker can:
-  read the sector info for proving from remote storage 
- miners can move the data from the current miner storage to the wdpost worker?

**Reply by siriusyim** [reply]
I understand that these two questions mean the same thing, no, can’t do this

**Reply by jennijuju** [reply]
could you please elaborate more on why it cant be done?

**Reply by magik6k** [reply]
> read the sector info for proving from remote storage

Sector info (the metadata) lives on the main mining node, not in sector storage, and it's sent to the PoSt worker when the mining node request PoSt to be computed. The PoSt worker will then just read challenges over http from other workers in the system.

(basically PoSt workers don't have any local storage, they just compute things, at least that's how I understand this design (and I like it being this way))

**Reply by siriusyim** [reply]
To get sector challenges, need to get the sector path information (pubSectorToPriv) first. If the sector is stored in the work, it will trigger the sector fetch (worker-->miner).

**Reply by siriusyim** [reply]
I will implement my ideas to solve this problem

**Comment by siriusyim** [comment]
Hi,@jennijuju We have started to submit PRs,
here is the PRs related to the algorithm：
https://github.com/filecoin-project/rust-fil-proofs/pull/1497
https://github.com/filecoin-project/rust-filecoin-proofs-api/pull/59
https://github.com/filecoin-project/filecoin-ffi/pull/197

**Comment by jennijuju** [comment]
@siriusmz ffi v10.1.0 is tagged and has the partition API you needed. Could you please start working on the lotus side of the changes and try to have a PR ready for review in mid Nov?

**Reply by jennijuju** [reply]
Ideally id like wdpost worker to be supported in lotus v1.13.2

**Reply by siriusyim** [reply]
I am revising and testing and will submit PR as soon as possible

**Comment by RobQuistNL** [comment, marked as answer]
@siriusmz are you still working on this? We could really use it :+)

**Reply by siriusyim** [reply]
The PR has been submitted, is being reviewed and communicated
filecoin-project/filecoin-ffi#197

filecoin-project/lotus/pull/7600

**Reply by RobQuistNL** [reply]
awesome, thanks!
